# ARG_neural

This repository contains a two-stage pipeline for classifying Antibiotic Resistance Genes (ARGs):

- **ARG-BERT_DNN**: Uses a pre-trained ProteinBERT model to generate contextual embeddings from protein sequences.
- **ARGneural**: Performs final ARG classification using the embeddings generated by ARG-BERT_DNN, along with optional auxiliary information such as drug types and species.

The **ARG-BERT_DNN** module is developed and tested using **Python 3.8** and **PyTorch 2.0.1**.

The **ARGneural** module is developed and tested using **Python 3.11** and **PyTorch 2.0.1**.

---

## How to Run

### 1. Clone the repository

```bash
git clone https://github.com/moriken814/ARG_neural.git
cd ARG_neural
```

### 2. Dataset
The repository assumes input sequences are formatted in CSV files. An example can be found in ARG-BERT_DNN/Sample_data/input.csv.

Obtain data from the HMD-ARG database: http://www.cbrc.kaust.edu.sa/HMDARG/dataset.html, and generate the required input files according to the procedure described in the associated paper.

For standard datasets, place the files under:
`ARG-BERT_DNN/inputs/HMDARG-DB/`
with the following file names:

`fold_{FOLD_NUM}.train.csv`

`fold_{FOLD_NUM}.test.csv`

For the Low Homology Dataset, place the files under:
`ARG-BERT_DNN/inputs/LHD/c{THRESHOLD}/`
with the following file names:

`fold_{FOLD_NUM}_{THRESHOLD}.train.csv`

`fold_{FOLD_NUM}_{THRESHOLD}.test.csv`


### 3. Run
**Step 1: Generate protein embeddings using ARG-BERT**

To use the Standard Dataset:
```bash
cd ARG-BERT_DNN
python train.py -f <fold> -s <seed>
```
To use the Low Homology Dataset:
```bash
python train.py -f <fold> -LHD -t <threshold> -s <seed>
```

**Step 2: Predict ARG classes using the neural network**

```bash
cd ARGneural
python run_multiclass.py -e <epoch num> -b <batch size>
```
In `run_multiclass.py`, you need to manually specify which dataset to use within the script.
